compiler/resolution/postFold.cpp

```diff
@@ -375,20 +375,40 @@ static Expr* postFoldPrimop(CallExpr* call) {
       call->replace(retval);
     }
 
-  } else if (call->isPrimitive("string_length") == true) {
+  } else if (call->isPrimitive("string_length_bytes") == true) {
     SymExpr* se = toSymExpr(call->get(1));
 
     INT_ASSERT(se);
 
     if (se->symbol()->isParameter() == true) {
-      const char* str    = get_string(se);
-      int         length = unescapeString(str, se).length();
+      const char* str     = get_string(se);
+      const size_t nbytes = unescapeString(str, se).length();
 
-      retval = new SymExpr(new_IntSymbol(length, INT_SIZE_DEFAULT));
+      retval = new SymExpr(new_IntSymbol(nbytes, INT_SIZE_DEFAULT));
 
       call->replace(retval);
     }
 
+  } else if (call->isPrimitive("string_length_codepoints") == true) {
+    SymExpr* se = toSymExpr(call->get(1));
+
+    INT_ASSERT(se && se->symbol()->isParameter());
+
+    const char* str         = get_string(se);
+    const std::string unesc = unescapeString(str, se);
+    const size_t nbytes     = unesc.length();
+
+    // Don't bother looking at the first byte.
+    // Count it as an initial UTF-8 byte.
```

Comment:
Why not look at the first byte? If this is an optimization I really doubt it will have meaningful performance impact...

Reply:
> Why not look at the first byte?

It's actually defensive, for now.  If the programmer somehow manages to start the string with an invalid byte, I don't want it to disrupt the compiler.  This also matches the behavior of the non-param `string.numCodepoints`.

Longer term, in the deep dive slides I've mentioned that when to verify that a string is valid UTF-8 is one of the two remaining decisions to be made.